modules/schema/src/main/java/org/apache/ignite/internal/schema/registry/UpgradingRowAdapter.java

```diff
@@ -31,21 +36,339 @@
     /** Column mapper. */
     private final ColumnMapper mapper;
 
+    /** Adapter schema. */
+    private final SchemaDescriptor schema;
+
     /**
-     * @param schema Schema descriptor of new version.
+     * @param schema Row adapter schema descriptor.
+     * @param rowSchema Row schema descriptor.
      * @param row Row.
      * @param mapper Column mapper.
      */
-    UpgradingRowAdapter(SchemaDescriptor schema, BinaryRow row, ColumnMapper mapper) {
-        super(schema, row);
+    UpgradingRowAdapter(SchemaDescriptor schema, SchemaDescriptor rowSchema, BinaryRow row, ColumnMapper mapper) {
+        super(rowSchema, row);
 
+        this.schema = schema;
         this.mapper = mapper;
     }
 
     /** {@inheritDoc} */
-    @Override protected long findColumn(int colIdx, NativeTypeSpec type) throws InvalidTypeException {
-        int mapIdx = mapper.map(colIdx);
+    @Override public SchemaDescriptor rowSchema() {
+        return schema;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int schemaVersion() {
+        return schema.version();
+    }
+
+    /**
+     * Map column.
+     *
+     * @param colIdx Column index in source schema.
+     * @return Column index in targer schema.
+     */
+    private int mapColumn(int colIdx) throws InvalidTypeException {
+        return mapper.map(colIdx);
+    }
+
+    /**
+     * Reads value for specified column.
+     *
+     * @param colIdx Column index.
+     * @return Column value.
+     * @throws InvalidTypeException If actual column type does not match the requested column type.
+     */
+    @Override public byte byteValue(int colIdx) throws InvalidTypeException {
+        int mappedId = mapColumn(colIdx);
+
+        Column column = mappedId < 0 ? mapper.mappedColumn(colIdx) : super.rowSchema().column(mappedId);
```

Comment:
looks like this could be simplified to `Column column = mapper.mappedColumn(colIdx)`